Mathlib/Analysis/Normed/Group/Seminorm.lean

```diff
@@ -0,0 +1,1020 @@
+
+theorem lt_def : p < q ↔ (p : E → ℝ) < q :=
+  Iff.rfl
+#align nonarch_add_group_seminorm.lt_def NonarchAddGroupSeminorm.lt_def
+
+@[simp, norm_cast]
+theorem coe_le_coe : (p : E → ℝ) ≤ q ↔ p ≤ q :=
+  Iff.rfl
+#align nonarch_add_group_seminorm.coe_le_coe NonarchAddGroupSeminorm.coe_le_coe
+
+@[simp, norm_cast]
+theorem coe_lt_coe : (p : E → ℝ) < q ↔ p < q :=
+  Iff.rfl
+#align nonarch_add_group_seminorm.coe_lt_coe NonarchAddGroupSeminorm.coe_lt_coe
+
+variable (p q) (f : F →+ E)
+
+instance : Zero (NonarchAddGroupSeminorm E) :=
+  ⟨{  toFun := 0
+      map_zero' := Pi.zero_apply _
+      add_le_max' := fun r s => by simp only [Pi.zero_apply]; rw [max_eq_right]; rfl
+      neg' := fun x => rfl }⟩
+
+@[simp, norm_cast]
+theorem coe_zero : ⇑(0 : NonarchAddGroupSeminorm E) = 0 :=
+  rfl
+#align nonarch_add_group_seminorm.coe_zero NonarchAddGroupSeminorm.coe_zero
+
+@[simp]
+theorem zero_apply (x : E) : (0 : NonarchAddGroupSeminorm E) x = 0 :=
+  rfl
+#align nonarch_add_group_seminorm.zero_apply NonarchAddGroupSeminorm.zero_apply
+
+instance : Inhabited (NonarchAddGroupSeminorm E) :=
+  ⟨0⟩
+
+-- TODO: define `SupSet` too, from the skeleton at
+-- https://github.com/leanprover-community/mathlib/pull/11329#issuecomment-1008915345
+instance : HasSup (NonarchAddGroupSeminorm E) :=
+  ⟨fun p q =>
+    { toFun := p ⊔ q
+      map_zero' := by rw [Pi.sup_apply, ← map_zero p, sup_eq_left, map_zero p, map_zero q]
+      add_le_max' := fun x y =>
+        sup_le ((map_add_le_max p x y).trans <| max_le_max le_sup_left le_sup_left)
+          ((map_add_le_max q x y).trans <| max_le_max le_sup_right le_sup_right)
+      neg' := fun x => by simp_rw [Pi.sup_apply, Pi.sup_apply, map_neg_eq_map p, map_neg_eq_map q]}⟩
+
+@[simp, norm_cast]
+theorem coe_sup : ⇑(p ⊔ q) = ⇑p ⊔ ⇑q :=
+  rfl
+#align nonarch_add_group_seminorm.coe_sup NonarchAddGroupSeminorm.coe_sup
+
+@[simp]
+theorem sup_apply (x : E) : (p ⊔ q) x = p x ⊔ q x :=
+  rfl
+#align nonarch_add_group_seminorm.sup_apply NonarchAddGroupSeminorm.sup_apply
+
+noncomputable instance : SemilatticeSup (NonarchAddGroupSeminorm E) :=
+  FunLike.coe_injective.semilatticeSup _ coe_sup
+
+end AddGroup
+
+section AddCommGroup
+
+variable [AddCommGroup E] [AddCommGroup F] (p q : NonarchAddGroupSeminorm E) (x y : E)
+
+theorem add_bddBelow_range_add {p q : NonarchAddGroupSeminorm E} {x : E} :
+    BddBelow (range fun y => p y + q (x - y)) :=
+  ⟨0, by
+    rintro _ ⟨x, rfl⟩
+    dsimp
+    exact add_nonneg (map_nonneg _ _) (map_nonneg _ _)⟩
+    -- porting note: was `positivity`
+#align nonarch_add_group_seminorm.add_bdd_below_range_add NonarchAddGroupSeminorm.add_bddBelow_range_add
+
+end AddCommGroup
+
+end NonarchAddGroupSeminorm
+
+namespace GroupSeminorm
+
+variable [Group E] [SMul R ℝ] [SMul R ℝ≥0] [IsScalarTower R ℝ≥0 ℝ]
+
+-- porting note: added manually because of problems with `to_additive`
+instance _root_.AddGroupSeminorm.toOne [DecidableEq E] [AddGroup E] : One (AddGroupSeminorm E) :=
```

Comment:
This is the same instance as on line 452, right?

Reply:
whoops, thank for spotting that. They were so far away from each other that I didn't notice.